### PR TITLE
Reland "[TypeProf][InstrPGO] Introduce raw and instr profile format change for type profiling."

### PR DESCRIPTION
**Pull request by mingmingl-llvm**
New change on top of [reviewed patch](https://github.com/llvm/llvm-project/pull/81691) are [in commits after this one](https://github.com/llvm/llvm-project/pull/82711/commits/d0757f46b3e3865b5f7c552bc0744309a363e0ac). Previous commits are restored from the remote branch with timestamps.

1. Fix build breakage for non-ELF platforms, by defining the missing functions {`__llvm_profile_begin_vtables`, `__llvm_profile_end_vtables`, `__llvm_profile_begin_vtabnames `, `__llvm_profile_end_vtabnames`} everywhere.
   * Tested on mac laptop (for darwins) and Windows. Specifically,  functions in `InstrProfilingPlatformWindows.c` returns `NULL` to make it more explicit that type prof isn't supported; see comments for the reason.
    * For the rest (AIX, other), mostly follow existing examples (like this [one](https://github.com/llvm/llvm-project/commit/f95b2f1acf1171abb0d00089fd4c9238753847e3))
   
2. Rename `__llvm_prf_vtabnames` -> `__llvm_prf_vns` for shorter section name, and make returned pointers [const](https://github.com/llvm/llvm-project/pull/82711/commits/a825d2a4ec00f07772a373091a702f149c3b0c34#diff-4de780ce726d76b7abc9d3353aef95013e7b21e7bda01be8940cc6574fb0b5ffR120-R121) 

**Original Description**

* Raw profile format
   - Header: records the byte size of compressed vtable names, and the number of profiled vtable entries (call it `VTableProfData`). Header also records padded bytes of each section.
   - Payload: adds a section for compressed vtable names, and a section to store `VTableProfData`. Both sections are padded so the size is a multiple of 8.
* Indexed profile format
  - Header: records the byte offset of compressed vtable names.
  - Payload: adds a section to store compressed vtable names. This section is used by `llvm-profdata` to show the list of vtables profiled for an instrumented site.
  
[The originally reviewed patch](https://github.com/llvm/llvm-project/pull/66825) will have profile reader/write change and llvm-profdata change.
- To ensure this PR has all the necessary profile format change along with profile version bump, created a copy of the originally reviewed patch in https://github.com/llvm/llvm-project/pull/80761. The copy doesn't have profile format change, but it has the set of tests which covers type profile generation, profile read and profile merge. Tests pass there.
  
rfc in https://discourse.llvm.org/t/rfc-dynamic-type-profiling-and-optimizations-in-llvm/74600

### CODE REVIEW

**Comment by minglotus-6** [comment]
This is a link (https://lab.llvm.org/buildbot/#/builders/127/builds/62532/steps/8/logs/stdio) for previously broken builds.

File: compiler-rt/lib/profile/InstrProfilingPlatformWindows.c [hunk 1]
```diff
@@ -59,9 +61,17 @@ const __llvm_profile_data *__llvm_profile_begin_data(void) {
...
 
+// TODO: Implement the linker magic on Windows.
+const VTableProfData *__llvm_profile_begin_vtables(void) { return NULL; }
+
+const VTableProfData *__llvm_profile_end_vtables(void) { return NULL; }
+
 const char *__llvm_profile_begin_names(void) { return &NamesStart + 1; }
 const char *__llvm_profile_end_names(void) { return &NamesEnd; }
 
+const char *__llvm_profile_begin_vtabnames(void) { return NULL; }
```
Comment (minglotus-6) [inline compiler-rt/lib/profile/InstrProfilingPlatformWindows.c:72]: https://gcc.godbolt.org/z/1Mv3qdjvK as a compile test.

**Comment by minglotus-6** [comment]
@w2yehia I wonder if you could help take a look at the changes to compiler-rt/lib/profile/InstrProfilingPlatformAIX.c? I expect them to be no-op in terms of new features, but would like a second pair of eye. Thanks!

**Review by w2yehia — COMMENTED** [review]
AIX changes lgtm.

**Comment by minglotus-6** [comment]
Update: I got the Windows machine. Windows linker magic [1] works out without compile errors or bad side effects, but the current type profiling won't work under MSVC ABI.  Thereby helper functions in `compiler-rt/lib/profile/InstrProfilingPlatformWindows.c` is made simple (just return `NULL`) such that it's very explicit type profiling isn't supported.

To elaborate on why current type profiling implementation doesn't work under MSVC ABI
* Under MSVC ABI, the name of the vtable variable is not necessarily the mangled name.  Take https://gcc.godbolt.org/z/drs1n9qd5 as an example, `@0`, `@1` and `@2` are vtables. 
* For instrumentation && profile collection under Itanium ABI, the vtables' mangled names (and its MD5hash) are used to map profiled runtime address to actual vtable variables.

As the godbolt example shows, the MSVC mangled name is still in the IR (just not vtable's variable name). I added some comments.

[1] diff in https://gist.github.com/minglotus-6/9df208e14f7c20b352cdb7f89d38c228

**Comment by whentojump** [comment]
Hi @minglotus-6 , can you please also update this place:

https://github.com/llvm/llvm-project/blob/82c6eeed08b1c8267f6e92d594c910fe57a9775e/llvm/include/llvm/ProfileData/InstrProf.h#L1183-L1198

And may I ask whether you plan to back port this to branch 18? (I'm asking because we have some dependency on this version and would like to know in advance). Thanks!

**Comment by minglotus-6** [comment]
> can you please also update this place

Thanks for the catch! Updated in https://github.com/llvm/llvm-project/commit/f87bde2962dded71599034880f57f742ea371888

> whether you plan to back port this to branch 18? (I'm asking because we have some dependency on this version and would like to know in advance)

Before seeing your comment, I don't have plans to back port the version update to LLVM version 18. Could you elaborate on how back porting a version bump to LLVM branch 18 helps? I have seen back-ports for bug fixes but haven't done it myself. I'm glad to back port if it helps.

**Comment by whentojump** [comment]
> Updated in https://github.com/llvm/llvm-project/commit/f87bde2962dded71599034880f57f742ea371888

Thanks!

By the way, https://llvm.org/docs/InstrProfileFormat.html is a really informative page. Thanks for your contributions!

> Before seeing your comment, I don't have plans to back port the version update to LLVM version 18. Could you elaborate on how back porting a version bump to LLVM branch 18 helps? I have seen back-ports for bug fixes but haven't done it myself. I'm glad to back port if it helps.

Sorry for not making it clear -- in fact, *not* backporting it can better help our use cases. :)) or :((

In short, we are instrumenting some OS kernels, like what [this patch](https://lore.kernel.org/linux-doc/20210407211704.367039-1-morbo@google.com/) does.

In kernel space, it's impossible to dump the profiles using any "runtime". So instead, some kernel code itself has to do this job. As a result, this piece of kernel code has to keep synchronized with LLVM code, e.g. in terms of profile format. That's why a stable format is certainly more desirable for us. You may notice in the linked patch: they were using raw profile version 5 three years ago and now we are 10 already. (In user space, this should not be a problem, because whatever produces or consumes the profile usually comes from the same version of LLVM.)

That being said, please make the decision at your own discretion, or based on broader community feedback. :))

**Comment by minglotus-6** [comment]
> in fact, not backporting it can better help our use cases. :)) or :((
>
> In short, we are instrumenting some OS kernels, like what [this patch](https://lore.kernel.org/linux-doc/20210407211704.367039-1-morbo@google.com/) does.
> ...
> That being said, please make the decision at your own discretion, or based on broader community feedback. :))

Ack, thanks for sharing!

**Comment by ayermolo** [comment]
Do you know by any chance how raw-{32, 64}-bits-{le, be}.test were generated?
I am trying to re-generate them with this patch, but without a [previous](https://github.com/llvm/llvm-project/commit/f95b2f1acf1171abb0d00089fd4c9238753847e3) changes, and having trouble getting 64 bit ones to pass.

**Comment by minglotus-6** [comment]
> Do you know by any chance how raw-{32, 64}-bits-{le, be}.test were generated?

I don't know how they were generated initially. When I updated these tests, I manually added the bytes in `RUN: printf` lines and initially had to run `llvm-profdata show` in a debugger (gdb) to get the test pass.

Given a binary raw profile, [hexdump](https://man7.org/linux/man-pages/man1/hexdump.1.html) should print the bytes ([example](https://github.com/llvm/llvm-project/compare/main...minglotus-6:llvm-project:endiannesstest#diff-1736a3ee25dde02bba55d670df78988fdb227e5a85b94b8707cf182cf70b28f0)). But the output is still painful to map bytes to the (likely multi-byte) fields in a (multi-section) raw profile.

> I am trying to re-generate them with this patch, but without a [previous](https://github.com/llvm/llvm-project/commit/f95b2f1acf1171abb0d00089fd4c9238753847e3) changes, and having trouble getting 64 bit ones to pass.

Is there a pointer showing the failed test? And which section (profile header, per function profile data, etc) is failing the raw profile reader?  I remember I got stumbled on updating [`CounterPtr`](https://github.com/llvm/llvm-project/blob/bcd586b5abfc7c22a1c0eca3a7abf18e7a09518d/compiler-rt/include/profile/InstrProfData.inc#L80) which records relative offset, illustrated with ASCII art in https://llvm.org/docs/InstrProfileFormat.html#profile-counters

**Comment by ayermolo** [comment]
The offset in to name section is wrong.
Thanks, will dig into hexdump output.

**Pull request merged**